mpcs/src/basefold/commit_phase.rs

```diff
@@ -201,54 +325,142 @@ where
+
+    // take codewrods match with target length then fold
+    let codewords_matched =
+        pop_front_while(codewords, |codeword| codeword.values.len() == target_len);
+    // take codewrods match next target length in preparation of being committed together
+    let codewords_next_level_matched = pop_front_while(codewords, |codeword| {
+        codeword.values.len() == next_level_target_len
+    });
+
+    // optimize for single codeword match
+    let folded_codeword = if (running_codeword_opt.is_some() as usize + codewords_matched.len())
+        == 1
+        && codewords_next_level_matched.is_empty()
+    {
+        RowMajorMatrix::new(
+            running_codeword_opt
+                .or_else(|| codewords_matched.first().map(|m| m.as_view()))
+                .unwrap()
+                .values
+                .par_chunks_exact(2)
+                .zip(folding_coeffs)
+                .map(|(ys, coeff)| {
+                    let (left, right) = (ys[0], ys[1]);
+                    // original (left, right) = (lo + hi*x, lo - hi*x), lo, hi are codeword, but after times x it's not codeword
+                    // recover left & right codeword via (lo, hi) = ((left + right) / 2, (left - right) / 2x)
+                    let (lo, hi) = ((left + right) * inv_2, (left - right) * *coeff); // e.g. coeff = (2 * dit_butterfly)^(-1) in rs code
+                    // we do fold on folded = (1-r) * left_codeword + r * right_codeword, as it match perfectly with raw message in lagrange domain fixed variable
+                    lo + (hi - lo) * challenge
+                })
+                .collect::<Vec<_>>(),
+            2,
+        )
+    } else {
+        RowMajorMatrix::new(
+            (0..target_len)
+                .into_par_iter()
+                .step_by(2)
+                .map(|index| {
+                    let coeff = &folding_coeffs[index >> 1];
+
+                    // 1st part folded with challenge then sum
+                    let cur_same_pos_sum = running_codeword_opt
+                        .into_iter()
+                        .chain(codewords_matched.iter().map(|m| m.as_view()))
+                        .map(|codeword| {
+                            let (left, right) =
+                                (codeword.values[index], codeword.values[index + 1]);
+                            // original (left, right) = (lo + hi*x, lo - hi*x), lo, hi are codeword, but after times x it's not codeword
+                            // recover left & right codeword via (lo, hi) = ((left + right) / 2, (left - right) / 2x)
+                            let (lo, hi) = ((left + right) * inv_2, (left - right) * *coeff); // e.g. coeff = (2 * dit_butterfly)^(-1) in rs code
+                            // we do fold on folded = (1-r) * left_codeword + r * right_codeword, as it match perfectly with raw message in lagrange domain fixed variable
+                            lo + (hi - lo) * challenge
```

Comment:
nitpick: we should abstract these lines of codes in a closure, as this can allow us to avoid the duplication.

Reply:
Good idea. Before widely refactor to define it in codeword, I define a util function and eliminate all of the duplicated logic in mpcs prover/verifier impl